openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -615,3 +617,216 @@ def test_user_change_with_no_ids(self):
+
+
+class TestEmailChangeMiddleware(TestSafeSessionsLogMixin, TestCase):
+    """
+    Test class for EmailChangeMiddleware
+    """
+
+    def setUp(self):
+        super().setUp()
+        self.TEST_PASSWORD = 'Password1234'
+        self.user = UserFactory.create(email='test@example.com', password=self.TEST_PASSWORD)
+        self.addCleanup(set_current_request, None)
+        self.request = get_mock_request(self.user)
+        self.request.session = {}
+        self.client.response = HttpResponse()
+        self.client.response.cookies = SimpleCookie()
+        self.addCleanup(RequestCache.clear_all_namespaces)
+
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=True)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    def test_process_request_user_not_authenticated(self, mock_mark_cookie_for_deletion):
+        """
+        Calls EmailChangeMiddleware.process_request when no user is authenticated.
+        Verifies that session and cookies are not affected.
+        """
+        # Unauthenticated User
+        self.request.user = AnonymousUser()
+
+        # Call process_request without authenticating a user
+        EmailChangeMiddleware(get_response=lambda request: None).process_request(self.request)
+
+        # Assert that session and cookies are not affected
+        # Assert that _mark_cookie_for_deletion not called
+        mock_mark_cookie_for_deletion.assert_not_called()
+
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=False)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_custom_attribute")
+    def test_process_request_toggle_disabled(self, mock_set_custom_attribute, mock_mark_cookie_for_deletion):
+        """
+        Calls EmailChangeMiddleware.process_request when user is authenticated but
+        ENFORCE_SESSION_EMAIL_MATCH is disabled.
+        Verifies that session and cookies are not affected.
+        """
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.TEST_PASSWORD)
+        self.request.session = self.client.session
+
+        # Ensure no email is stored in the session
+        self.assertIsNone(self.request.session.get('email'))
+
+        # Call process_request without authenticating a user
+        EmailChangeMiddleware(get_response=lambda request: None).process_request(self.request)
+
+        # Assert that session and cookies are not affected
+        # Assert that _mark_cookie_for_deletion not called
+        mock_mark_cookie_for_deletion.assert_not_called()
+
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=True)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_custom_attribute")
+    def test_process_request_emails_match(
```

Comment:
Maybe in addition to these tests, but I'm wondering if we could have some more resilient integration test, using the actual middleware, and possible an API call to actually update the user's email?
- In place of calls to `register_email_change`, `process_request` and `process_response`, you could instead make various requests after having set `MIDDLEWARE` appropriately, and ensure that the actual cookies have been deleted or updated, and the user has been logged out, etc.

See https://github.com/openedx/edx-drf-extensions/blob/9eea525bf25a5d2af2911a5e94f3026da649f9e8/edx_rest_framework_extensions/auth/jwt/tests/test_middleware.py#L523-L529 for the type of test I mean, although you probably won't need `ROOT_URLCONF` because you may be using real platform endpoints.

Since this is for security purposes, it would be great to know that all the pieces are working together. That said, I'd also timebox any effort to get this working. Hopefully it would be relatively simple.

Reply:
I am working on this one. Meanwhile please have a look the responses of other comments and let me know about your opinions on them.

Reply:
@robrap Thankyou. I will first give it a try and let you know.